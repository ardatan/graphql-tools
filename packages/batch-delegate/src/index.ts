export * from './batchDelegateToSchema.js';
export * from './createBatchDelegateFn.js';

export * from './types.js';
