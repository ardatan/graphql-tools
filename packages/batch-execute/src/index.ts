export * from './createBatchingExecutor.js';
export * from './getBatchingExecutor.js';
