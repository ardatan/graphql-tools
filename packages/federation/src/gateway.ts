import {
  buildASTSchema,
  concatAST,
  DocumentNode,
  FieldDefinitionNode,
  GraphQLSchema,
  Kind,
  ObjectTypeDefinitionNode,
  ObjectTypeExtensionNode,
  parse,
  visit,
} from 'graphql';
import { createDefaultExecutor, SubschemaConfig } from '@graphql-tools/delegate';
import { buildHTTPExecutor, HTTPExecutorOptions } from '@graphql-tools/executor-http';
import { stitchSchemas, SubschemaConfigTransform } from '@graphql-tools/stitch';
import {
  createGraphQLError,
  ExecutionResult,
  Executor,
  getDocumentNodeFromSchema,
  inspect,
} from '@graphql-tools/utils';
import { SubgraphBaseSDL } from './subgraph.js';
import {
  filterInternalFieldsAndTypes,
  getArgsFromKeysForFederation,
  getCacheKeyFnFromKey,
  getKeyForFederation,
} from './utils.js';

export const SubgraphSDLQuery = /* GraphQL */ `
  query SubgraphSDL {
    _service {
      sdl
    }
  }
`;

export async function getSubschemaForFederationWithURL(
  config: HTTPExecutorOptions,
): Promise<SubschemaConfig> {
  const executor = buildHTTPExecutor(config);
  const subschemaConfig = await getSubschemaForFederationWithExecutor(executor);
  return {
    batch: true,
    ...subschemaConfig,
  };
}

export function getSubschemaForFederationWithTypeDefs(typeDefs: DocumentNode): SubschemaConfig {
  const subschemaConfig = {} as SubschemaConfig;
  const typeMergingConfig = (subschemaConfig.merge = subschemaConfig.merge || {});
  const entityTypes: string[] = [];
  const visitor = (node: ObjectTypeDefinitionNode | ObjectTypeExtensionNode) => {
    if (node.directives) {
      const typeName = node.name.value;
      const selections: string[] = [];
      for (const directive of node.directives) {
        const directiveArgs = directive.arguments || [];
        switch (directive.name.value) {
          case 'key': {
            if (
              directiveArgs.some(
                arg =>
                  arg.name.value === 'resolvable' &&
                  arg.value.kind === Kind.BOOLEAN &&
                  arg.value.value === false,
              )
            ) {
              continue;
            }
            const selectionValueNode = directiveArgs.find(
              arg => arg.name.value === 'fields',
            )?.value;
            if (selectionValueNode?.kind === Kind.STRING) {
              selections.push(selectionValueNode.value);
            }
            break;
          }
          case 'inaccessible':
            return null;
        }
      }
      // If it is not an entity, continue
      if (selections.length === 0) {
        return node;
      }
      const typeMergingTypeConfig = (typeMergingConfig[typeName] =
        typeMergingConfig[typeName] || {});
      if (
        node.kind === Kind.OBJECT_TYPE_DEFINITION &&
        !node.directives?.some(d => d.name.value === 'extends')
      ) {
        typeMergingTypeConfig.canonical = true;
      }
      entityTypes.push(typeName);
      const selectionsStr = selections.join(' ');
      typeMergingTypeConfig.selectionSet = `{ ${selectionsStr} }`;
      typeMergingTypeConfig.dataLoaderOptions = {
        cacheKeyFn: getCacheKeyFnFromKey(selectionsStr),
      };
      typeMergingTypeConfig.argsFromKeys = getArgsFromKeysForFederation;
      typeMergingTypeConfig.fieldName = `_entities`;
      typeMergingTypeConfig.key = getKeyForFederation;
      const fields = [];
      if (node.fields) {
        for (const fieldNode of node.fields) {
          let removed = false;
          if (fieldNode.directives) {
            const fieldName = fieldNode.name.value;
            for (const directive of fieldNode.directives) {
              const directiveArgs = directive.arguments || [];
              switch (directive.name.value) {
                case 'requires': {
                  const typeMergingFieldsConfig = (typeMergingTypeConfig.fields =
                    typeMergingTypeConfig.fields || {});
                  typeMergingFieldsConfig[fieldName] = typeMergingFieldsConfig[fieldName] || {};
                  if (
                    directiveArgs.some(
                      arg =>
                        arg.name.value === 'resolvable' &&
                        arg.value.kind === Kind.BOOLEAN &&
                        arg.value.value === false,
                    )
                  ) {
                    continue;
                  }
                  const selectionValueNode = directiveArgs.find(
                    arg => arg.name.value === 'fields',
                  )?.value;
                  if (selectionValueNode?.kind === Kind.STRING) {
                    typeMergingFieldsConfig[fieldName].selectionSet =
                      `{ ${selectionValueNode.value} }`;
                    typeMergingFieldsConfig[fieldName].computed = true;
                  }
                  break;
                }
                case 'external':
                case 'inaccessible': {
                  removed = !typeMergingTypeConfig.selectionSet?.includes(` ${fieldName} `);
                  break;
                }
                case 'override': {
                  const typeMergingFieldsConfig = (typeMergingTypeConfig.fields =
                    typeMergingTypeConfig.fields || {});
                  typeMergingFieldsConfig[fieldName] = typeMergingFieldsConfig[fieldName] || {};
                  typeMergingFieldsConfig[fieldName].canonical = true;
                  break;
                }
              }
            }
          }
          if (!removed) {
            fields.push(fieldNode);
          }
        }
        (node.fields as FieldDefinitionNode[]) = fields;
      }
    }
    return {
      ...node,
      kind: Kind.OBJECT_TYPE_DEFINITION,
    };
  };
  const parsedSDL = visit(typeDefs, {
    ObjectTypeExtension: visitor,
    ObjectTypeDefinition: visitor,
  });
  let extraSdl = SubgraphBaseSDL;
  if (entityTypes.length > 0) {
    extraSdl += `\nunion _Entity = ${entityTypes.join(' | ')}`;
    extraSdl += `\nextend type Query { _entities(representations: [_Any!]!): [_Entity]! }`;
  }
  subschemaConfig.schema = buildASTSchema(concatAST([parse(extraSdl), parsedSDL]), {
    assumeValidSDL: true,
    assumeValid: true,
  });
  // subschemaConfig.batch = true;
  return subschemaConfig;
}

export async function getSubschemaForFederationWithExecutor(executor: Executor) {
  const sdlQueryResult = (await executor({
    document: parse(SubgraphSDLQuery),
  })) as ExecutionResult;

  if (sdlQueryResult.errors?.length) {
    const error = sdlQueryResult.errors[0];
    throw createGraphQLError(error.message, error);
  }

  if (!sdlQueryResult.data?._service?.sdl) {
    throw new Error(`Unexpected result: ${inspect(sdlQueryResult)}`);
  }

  const typeDefs = parse(sdlQueryResult.data._service.sdl);

  const subschemaConfig = getSubschemaForFederationWithTypeDefs(typeDefs);

  return {
    ...subschemaConfig,
    executor,
  };
}

export async function getSubschemaForFederationWithSchema(schema: GraphQLSchema) {
  const executor = createDefaultExecutor(schema);
  return getSubschemaForFederationWithExecutor(executor);
}

export async function getStitchedSchemaWithUrls(configs: HTTPExecutorOptions[]) {
  const subschemas = await Promise.all(
    configs.map(config => getSubschemaForFederationWithURL(config)),
  );
  const schema = stitchSchemas({
    subschemas,
  });
  return filterInternalFieldsAndTypes(schema);
}

export const federationSubschemaTransformer: SubschemaConfigTransform =
  function federationSubschemaTransformer(subschemaConfig: SubschemaConfig): SubschemaConfig {
    const typeDefs = getDocumentNodeFromSchema(subschemaConfig.schema);
    const newSubschema = getSubschemaForFederationWithTypeDefs(typeDefs);
    return {
      ...subschemaConfig,
      ...newSubschema,
      merge: {
        ...newSubschema.merge,
        ...subschemaConfig.merge,
      },
    };
  };
