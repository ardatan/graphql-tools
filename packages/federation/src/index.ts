export * from './gateway.js';
export * from './subgraph.js';
export * from './supergraph.js';
export * from './utils.js';
