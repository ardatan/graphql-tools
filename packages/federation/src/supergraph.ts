import {
  buildASTSchema,
  DocumentNode,
  EnumTypeDefinitionNode,
  EnumValueDefinitionNode,
  FieldDefinitionNode,
  Kind,
  NamedTypeNode,
  ObjectTypeDefinitionNode,
  parse,
  ScalarTypeDefinitionNode,
  TypeDefinitionNode,
  UnionTypeDefinitionNode,
  visit,
} from 'graphql';
import { MergedTypeConfig, SubschemaConfig } from '@graphql-tools/delegate';
import { buildHTTPExecutor } from '@graphql-tools/executor-http';
import { stitchSchemas } from '@graphql-tools/stitch';
import type { Executor } from '@graphql-tools/utils';
import {
  filterInternalFieldsAndTypes,
  getArgsFromKeysForFederation,
  getKeyForFederation,
  getNamedTypeNode,
} from './utils.js';

export interface GetSubschemasFromSupergraphSdlOpts {
  supergraphSdl: string | DocumentNode;
  onExecutor?: (opts: { subgraphName: string; endpoint: string }) => Executor;
  batch?: boolean;
}

export function getSubschemasFromSupergraphSdl({
  supergraphSdl,
  onExecutor = ({ endpoint }) => buildHTTPExecutor({ endpoint }),
  batch = false,
}: GetSubschemasFromSupergraphSdlOpts) {
  const ast =
    typeof supergraphSdl === 'string' ? parse(supergraphSdl, { noLocation: true }) : supergraphSdl;
  const subgraphQueryFieldDefinitionNodes = new Map<string, FieldDefinitionNode[]>();
  const subgraphEndpointMap = new Map<string, string>();
  const subgraphTypesMap = new Map<string, TypeDefinitionNode[]>();
  const typeNameKeyBySubgraphMap = new Map<string, Map<string, string>>();
  const typeNameFieldsKeyBySubgraphMap = new Map<string, Map<string, Map<string, string>>>();
  const typeNameCanonicalMap = new Map<string, string>();
  const unownedTypes = new Map<string, TypeDefinitionNode>();
  const dependenciesBySubgraphMap = new Map<string, Set<string>>();
  visit(ast, {
    ScalarTypeDefinition(node) {
      unownedTypes.set(node.name.value, node);
    },
    InterfaceTypeDefinition(node) {
      unownedTypes.set(node.name.value, node);
    },
    UnionTypeDefinition(node) {
      unownedTypes.set(node.name.value, node);
    },
    EnumTypeDefinition(node) {
      if (node.name.value === 'join__Graph') {
        node.values?.forEach(valueNode => {
          const joinGraphDirectiveNode = valueNode.directives?.find(
            directiveNode => directiveNode.name.value === 'join__graph',
          );
          if (joinGraphDirectiveNode) {
            const urlArgumentNode = joinGraphDirectiveNode.arguments?.find(
              argumentNode => argumentNode.name.value === 'url',
            );
            if (urlArgumentNode?.value?.kind === Kind.STRING) {
              subgraphEndpointMap.set(valueNode.name.value, urlArgumentNode.value.value);
            }
          }
        });
      }
      node.directives?.forEach(directiveNode => {
        if (directiveNode.name.value === 'join__type') {
          const graphArgumentNode = directiveNode.arguments?.find(
            argumentNode => argumentNode.name.value === 'graph',
          );
          if (graphArgumentNode?.value?.kind === Kind.ENUM) {
            const graphName = graphArgumentNode.value.value;
            const enumValueNodes: EnumValueDefinitionNode[] =
              node.values?.filter(valueNode => {
                const joinEnumValueDirectiveNode = valueNode.directives?.find(
                  directiveNode => directiveNode.name.value === 'join__enumValue',
                );
                if (joinEnumValueDirectiveNode) {
                  const graphArgumentNode = joinEnumValueDirectiveNode.arguments?.find(
                    argumentNode => argumentNode.name.value === 'graph',
                  );
                  if (graphArgumentNode?.value?.kind === Kind.ENUM) {
                    return graphArgumentNode.value.value === graphName;
                  }
                }
                return false;
              }) || [];
            const enumTypedDefNodeForSubgraph: EnumTypeDefinitionNode = {
              ...node,
              values: enumValueNodes,
            };
            let subgraphTypes = subgraphTypesMap.get(graphName);
            if (!subgraphTypes) {
              subgraphTypes = [];
              subgraphTypesMap.set(graphName, subgraphTypes);
            }
            subgraphTypes.push(enumTypedDefNodeForSubgraph);
          }
        }
      });
    },
    ObjectTypeDefinition(node) {
      if (node.name.value === 'Query') {
        node.fields?.forEach(fieldNode => {
          fieldNode.directives?.forEach(directiveNode => {
            if (directiveNode.name.value === 'join__field') {
              const graphArgumentNode = directiveNode.arguments?.find(
                argumentNode => argumentNode.name.value === 'graph',
              );
              if (graphArgumentNode?.value?.kind === Kind.ENUM) {
                const graphName = graphArgumentNode.value.value;
                let fieldDefinitionNodesOfSubgraph =
                  subgraphQueryFieldDefinitionNodes.get(graphName);
                if (!fieldDefinitionNodesOfSubgraph) {
                  fieldDefinitionNodesOfSubgraph = [];
                  subgraphQueryFieldDefinitionNodes.set(graphName, fieldDefinitionNodesOfSubgraph);
                }
                fieldDefinitionNodesOfSubgraph.push(fieldNode);
              }
            }
          });
        });
      }
      let owned = false;
      node.directives?.forEach(directiveNode => {
        if (directiveNode.name.value === 'join__owner') {
          const graphArgumentNode = directiveNode.arguments?.find(
            argumentNode => argumentNode.name.value === 'graph',
          );
          if (graphArgumentNode?.value?.kind === Kind.ENUM) {
            typeNameCanonicalMap.set(node.name.value, graphArgumentNode.value.value);
          }
        }
        if (directiveNode.name.value === 'join__type') {
          const graphArgumentNode = directiveNode.arguments?.find(
            argumentNode => argumentNode.name.value === 'graph',
          );
          if (graphArgumentNode?.value?.kind === Kind.ENUM) {
            let keyFieldName: string;
            const graphName = graphArgumentNode.value.value;
            const keyArgumentNode = directiveNode.arguments?.find(
              argumentNode => argumentNode.name.value === 'key',
            );
            if (keyArgumentNode?.value?.kind === Kind.STRING) {
              let typeNameKeyMap = typeNameKeyBySubgraphMap.get(graphName);
              if (!typeNameKeyMap) {
                typeNameKeyMap = new Map();
                typeNameKeyBySubgraphMap.set(graphName, typeNameKeyMap);
              }
              keyFieldName = keyArgumentNode.value.value;
              typeNameKeyMap.set(node.name.value, keyFieldName);
            }
            const fieldDefinitionNodesOfSubgraph =
              node.fields?.filter(fieldNode => {
                if (fieldNode.name.value === keyFieldName) {
                  return true;
                }
                const joinFieldDirectiveNode = fieldNode.directives?.find(
                  directiveNode => directiveNode.name.value === 'join__field',
                );
                if (joinFieldDirectiveNode) {
                  const graphArgumentNode = joinFieldDirectiveNode.arguments?.find(
                    argumentNode => argumentNode.name.value === 'graph',
                  );
                  if (graphArgumentNode?.value?.kind === Kind.ENUM) {
                    if (graphArgumentNode.value.value === graphName) {
                      let dependenciesBySubgraph = dependenciesBySubgraphMap.get(graphName);
                      if (!dependenciesBySubgraph) {
                        dependenciesBySubgraph = new Set();
                        dependenciesBySubgraphMap.set(graphName, dependenciesBySubgraph);
                      }
                      const namedTypeNode = getNamedTypeNode(fieldNode.type);
                      dependenciesBySubgraph.add(namedTypeNode.name.value);

                      const requiresArgumentNode = joinFieldDirectiveNode.arguments?.find(
                        argumentNode => argumentNode.name.value === 'requires',
                      );
                      if (requiresArgumentNode?.value?.kind === Kind.STRING) {
                        let typeNameFieldsKeyMap = typeNameFieldsKeyBySubgraphMap.get(graphName);
                        if (!typeNameFieldsKeyMap) {
                          typeNameFieldsKeyMap = new Map();
                          typeNameFieldsKeyBySubgraphMap.set(graphName, typeNameFieldsKeyMap);
                        }
                        let fieldsKeyMap = typeNameFieldsKeyMap.get(node.name.value);
                        if (!fieldsKeyMap) {
                          fieldsKeyMap = new Map();
                          typeNameFieldsKeyMap.set(node.name.value, fieldsKeyMap);
                        }
                        fieldsKeyMap.set(fieldNode.name.value, requiresArgumentNode.value.value);
                      }
                      return true;
                    }
                  }
                } else {
                  return true;
                }
                return false;
              }) || [];
            const objectTypedDefNodeForSubgraph: ObjectTypeDefinitionNode = {
              ...node,
              fields: fieldDefinitionNodesOfSubgraph,
            };
            let subgraphTypes = subgraphTypesMap.get(graphName);
            if (!subgraphTypes) {
              subgraphTypes = [];
              subgraphTypesMap.set(graphName, subgraphTypes);
            }
            owned = true;
            subgraphTypes.push(objectTypedDefNodeForSubgraph);
          }
        }
      });
      if (!owned) {
        unownedTypes.set(node.name.value, node);
      }
    },
  });
  const subschemaMap = new Map<string, SubschemaConfig>();
  for (const [subgraphName, endpoint] of subgraphEndpointMap) {
    const executor = onExecutor({ subgraphName, endpoint });
    const mergeConfig: SubschemaConfig['merge'] = {};
    const typeNameKeyMap = typeNameKeyBySubgraphMap.get(subgraphName);
    const unionTypeNodes: NamedTypeNode[] = [];
    if (typeNameKeyMap) {
      const typeNameFieldsKeyMap = typeNameFieldsKeyBySubgraphMap.get(subgraphName);
      for (const [typeName, key] of typeNameKeyMap) {
        const mergedTypeConfig: MergedTypeConfig = (mergeConfig[typeName] = {
          selectionSet: `{ ${key} }`,
          argsFromKeys: getArgsFromKeysForFederation,
          key: getKeyForFederation,
          fieldName: `_entities`,
        });
        const keyProps = key.split(' ');
        mergedTypeConfig.dataLoaderOptions = {
          cacheKeyFn(root) {
            return keyProps.map(key => root[key]).join(' ');
          },
        };
        const fieldsKeyMap = typeNameFieldsKeyMap?.get(typeName);
        if (fieldsKeyMap) {
          const fieldsConfig = (mergedTypeConfig.fields = {});
          for (const [fieldName, key] of fieldsKeyMap) {
            fieldsConfig[fieldName] = {
              selectionSet: `{ ${key} }`,
              computed: true,
            };
          }
        }
        if (typeNameCanonicalMap.get(typeName) === subgraphName) {
          mergedTypeConfig.canonical = true;
        }
        unionTypeNodes.push({
          kind: Kind.NAMED_TYPE,
          name: {
            kind: Kind.NAME,
            value: typeName,
          },
        });
      }
    }
    const entitiesUnionTypeDefinitionNode: UnionTypeDefinitionNode = {
      name: {
        kind: Kind.NAME,
        value: '_Entity',
      },
      kind: Kind.UNION_TYPE_DEFINITION,
      types: unionTypeNodes,
    };
    const queryFields = subgraphQueryFieldDefinitionNodes.get(subgraphName) || [];
    const queryWithEntitiesFieldDefinitionNode: ObjectTypeDefinitionNode = {
      name: {
        kind: Kind.NAME,
        value: 'Query',
      },
      kind: Kind.OBJECT_TYPE_DEFINITION,
      fields: [
        ...queryFields,
        {
          kind: Kind.FIELD_DEFINITION,
          name: {
            kind: Kind.NAME,
            value: '_entities',
          },
          type: {
            kind: Kind.NAMED_TYPE,
            name: {
              kind: Kind.NAME,
              value: '_Entity',
            },
          },
          arguments: [
            {
              kind: Kind.INPUT_VALUE_DEFINITION,
              name: {
                kind: Kind.NAME,
                value: 'representations',
              },
              type: {
                kind: Kind.NON_NULL_TYPE,
                type: {
                  kind: Kind.LIST_TYPE,
                  type: {
                    kind: Kind.NON_NULL_TYPE,
                    type: {
                      kind: Kind.NAMED_TYPE,
                      name: {
                        kind: Kind.NAME,
                        value: '_Any',
                      },
                    },
                  },
                },
              },
            },
          ],
        },
      ],
    };
    const subgraphTypes = subgraphTypesMap.get(subgraphName) || [];
    const dependencyNames = dependenciesBySubgraphMap.get(subgraphName) || new Set();
    for (const dependencyName of dependencyNames) {
      const dependencyType = unownedTypes.get(dependencyName);
      if (dependencyType) {
        subgraphTypes.push(dependencyType);
      }
    }
    const schema = buildASTSchema(
      {
        kind: Kind.DOCUMENT,
        definitions: [
          ...subgraphTypes,
          entitiesUnionTypeDefinitionNode,
          anyTypeDefinitionNode,
          queryWithEntitiesFieldDefinitionNode,
        ],
      },
      {
        assumeValidSDL: true,
        assumeValid: true,
      },
    );
    subschemaMap.set(subgraphName, {
      schema,
      executor,
      merge: mergeConfig,
      batch,
    });
  }
  return subschemaMap;
}

export function getStitchedSchemaFromSupergraphSdl(opts: GetSubschemasFromSupergraphSdlOpts) {
  const subschemaMap = getSubschemasFromSupergraphSdl(opts);
  const supergraphSchema = stitchSchemas({
    subschemas: [...subschemaMap.values()],
  });
  return filterInternalFieldsAndTypes(supergraphSchema);
}

const anyTypeDefinitionNode: ScalarTypeDefinitionNode = {
  name: {
    kind: Kind.NAME,
    value: '_Any',
  },
  kind: Kind.SCALAR_TYPE_DEFINITION,
};
