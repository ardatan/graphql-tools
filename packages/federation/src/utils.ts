import { GraphQLSchema } from 'graphql';
import { FilterRootFields, FilterTypes } from '@graphql-tools/wrap';

export function getArgsFromKeysForFederation(representations: readonly any[]) {
  return { representations };
}

export function getKeyForFederation(root: any) {
  return root;
}

export function filterInternalFieldsAndTypes(finalSchema: GraphQLSchema) {
  const removeEntitiesField = new FilterRootFields(
    (operation, fieldName) =>
      !(operation === 'Query' && (fieldName === '_entities' || fieldName === '_sdl')),
  );
  const removeEntityAndAny = new FilterTypes(
    type =>
      type.name !== '_Entity' &&
      type.name !== '_Any' &&
      type.name !== '_FieldSet' &&
      type.name !== '_Service' &&
      !type.name.startsWith('link__'),
  );
  const fakeSubschemaConfig = {
    schema: finalSchema,
    transforms: [removeEntitiesField, removeEntityAndAny],
  };
  for (const transform of fakeSubschemaConfig.transforms) {
    finalSchema = transform.transformSchema(finalSchema, fakeSubschemaConfig);
  }
  return finalSchema;
}
