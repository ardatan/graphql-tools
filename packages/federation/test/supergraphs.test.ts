import { readdirSync, readFileSync } from 'fs';
import { join } from 'path';
import { buildSchema, lexicographicSortSchema, printSchema, versionInfo } from 'graphql';
import { filterSchema, getDirective, pruneSchema } from '@graphql-tools/utils';
import { getStitchedSchemaFromSupergraphSdl } from '../src/supergraph';

describe('Supergraphs', () => {
  if (versionInfo.major === 15) {
    it('skipping tests for graphql v15', () => {
      expect(true).toBe(true);
    });
    return;
  }
  readdirSync(join(__dirname, 'fixtures', 'supergraphs')).forEach(fixture => {
    describe(fixture, () => {
      const fixturePath = join(__dirname, 'fixtures', 'supergraphs', fixture);
      const supergraphSdl = readFileSync(fixturePath, 'utf8');
      it('matches', () => {
        const schema = getStitchedSchemaFromSupergraphSdl({ supergraphSdl });
        const sortedSchema = lexicographicSortSchema(schema);
        const sortedInputSchema = lexicographicSortSchema(
          buildSchema(supergraphSdl, { noLocation: true, assumeValid: true, assumeValidSDL: true }),
        );
        const filteredInputSchema = pruneSchema(
          filterSchema({
            schema: sortedInputSchema,
            typeFilter: typeName =>
              !typeName.startsWith('link__') &&
              !typeName.startsWith('join__') &&
              !typeName.startsWith('core__'),
            fieldFilter: (_, __, fieldConfig) =>
              !getDirective(sortedInputSchema, fieldConfig, 'inaccessible')?.length,
            directiveFilter: () => false,
            enumValueFilter: (_, __, enumValueConfig) =>
              !getDirective(sortedInputSchema, enumValueConfig, 'inaccessible')?.length,
          }),
        );
        expect(printSchema(pruneSchema(sortedSchema)).trim()).toBe(
          printSchema(filteredInputSchema).trim(),
        );
      });
    });
  });
});
