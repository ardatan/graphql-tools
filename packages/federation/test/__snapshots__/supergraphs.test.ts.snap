// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Supergraphs a.graphql matches: a.graphql - stitchedSchema 1`] = `
"schema {
  query: Query
}

type Query {
  me: User
  users: [User]
  topProducts(first: Int): [Product]
}

type User {
  numberOfReviews: Int
  reviews: [Review]
  username: String
  birthDate: String
  id: ID!
  name: String
}

type Product {
  reviews(input: ProductReviewsInput): [Review]
  inStock: Boolean
  shippingEstimate: Int
  name: String
  price: Int
  upc: String!
  weight: Int
}

enum ReviewsOrder {
  ASC
  DESC
}

input ProductReviewsInput {
  first: Int
  order: ReviewsOrder
}

type AnonymousReview implements Review {
  body: String
  id: ID!
  product: Product
}

interface Review {
  body: String
  id: ID!
  product: Product
}

scalar Stars

type UserReview implements Review {
  author: User
  body: String
  id: ID!
  product: Product
  stars: Stars
}"
`;

exports[`Supergraphs a.graphql subgraphs: a.graphql - SERVICE0 1`] = `
"schema {
  query: Query
}

type Query {
  me: User
  users: [User]
  _entities(representations: [_Any!]!): _Entity
}

type User {
  birthDate: String
  id: ID!
  name: String
  username: String
}

union _Entity = User

scalar _Any"
`;

exports[`Supergraphs a.graphql subgraphs: a.graphql - SERVICE1 1`] = `
"schema {
  query: Query
}

type Product {
  name: String
  price: Int
  upc: String!
  weight: Int
}

type Query {
  topProducts(first: Int): [Product]
  _entities(representations: [_Any!]!): _Entity
}

union _Entity = Product

scalar _Any"
`;

exports[`Supergraphs a.graphql subgraphs: a.graphql - SERVICE2 1`] = `
"schema {
  query: Query
}

enum ReviewsOrder {
  ASC
  DESC
}

input ProductReviewsInput {
  first: Int
  order: ReviewsOrder
}

type AnonymousReview implements Review {
  body: String
  id: ID!
  product: Product
}

type Product {
  reviews(input: ProductReviewsInput): [Review]
}

type Query {
  _entities(representations: [_Any!]!): _Entity
}

interface Review {
  body: String
  id: ID!
  product: Product
}

scalar Stars

type User {
  numberOfReviews: Int
  reviews: [Review]
  username: String
}

type UserReview implements Review {
  author: User
  body: String
  id: ID!
  product: Product
  stars: Stars
}

union _Entity = AnonymousReview | Product | User | UserReview

scalar _Any"
`;

exports[`Supergraphs a.graphql subgraphs: a.graphql - SERVICE3 1`] = `
"schema {
  query: Query
}

type Product {
  inStock: Boolean
  shippingEstimate: Int
}

type Query {
  _entities(representations: [_Any!]!): _Entity
}

union _Entity = Product

scalar _Any"
`;

exports[`Supergraphs b.graphql matches: b.graphql - stitchedSchema 1`] = `
"schema {
  query: Query
}

type Query {
  allPandas: [Panda]
  panda(name: ID!): Panda
  allProducts: [ProductItf]
  product(id: ID!): ProductItf
  review(id: Int!): Review
}

type DeliveryEstimates {
  estimatedDelivery: String
  fastestDelivery: String
}

type Product implements ProductItf & SkuItf {
  id: ID! @tag(name: "hi-from-products")
  delivery(zip: String): DeliveryEstimates
  dimensions: ProductDimension
  sku: String
  name: String @hello
  package: String
  variation: ProductVariation
  createdBy: User
  hidden: String
  oldField: String
  reviewsScore: Float!
  reviewsCount: Int!
  reviews: [Review!]!
}

type ProductDimension {
  size: String @hello
  weight: Float
}

interface ProductItf implements SkuItf {
  id: ID!
  dimensions: ProductDimension
  delivery(zip: String): DeliveryEstimates
  sku: String
  name: String
  package: String
  variation: ProductVariation
  createdBy: User
  oldField: String @deprecated(reason: "refactored out")
  reviewsCount: Int!
  reviewsScore: Float!
  reviews: [Review!]!
}

enum ShippingClass {
  STANDARD
  EXPRESS
  OVERNIGHT
}

type Panda {
  name: ID!
  favoriteFood: String @tag(name: "nom-nom-nom")
}

type ProductVariation {
  id: ID!
  name: String
}

interface SkuItf {
  sku: String
}

type User {
  email: ID! @tag(name: "test-from-users")
  totalProductsCreated: Int
  name: String
}

type Review {
  id: Int!
  body: String!
}"
`;

exports[`Supergraphs b.graphql subgraphs: b.graphql - INVENTORY 1`] = `
"schema {
  query: Query
}

type DeliveryEstimates {
  estimatedDelivery: String
  fastestDelivery: String
}

type Product implements ProductItf {
  id: ID! @tag(name: "hi-from-products")
  delivery(zip: String): DeliveryEstimates
}

type ProductDimension {
  size: String @hello
  weight: Float
}

interface ProductItf {
  id: ID!
  dimensions: ProductDimension
  delivery(zip: String): DeliveryEstimates
}

type Query {
  _entities(representations: [_Any!]!): _Entity
}

enum ShippingClass {
  STANDARD
  EXPRESS
  OVERNIGHT
}

union _Entity = Product

scalar _Any"
`;

exports[`Supergraphs b.graphql subgraphs: b.graphql - PANDAS 1`] = `
"schema {
  query: Query
}

type Panda {
  name: ID!
  favoriteFood: String @tag(name: "nom-nom-nom")
}

type Query {
  allPandas: [Panda]
  panda(name: ID!): Panda
  _entities(representations: [_Any!]!): _Entity
}

union _Entity

scalar _Any"
`;

exports[`Supergraphs b.graphql subgraphs: b.graphql - PRODUCTS 1`] = `
"schema {
  query: Query
}

type Product implements ProductItf & SkuItf {
  id: ID! @tag(name: "hi-from-products")
  dimensions: ProductDimension
  sku: String
  name: String @hello
  package: String
  variation: ProductVariation
  createdBy: User
  hidden: String
  oldField: String
}

type ProductDimension {
  size: String @hello
  weight: Float
}

interface ProductItf implements SkuItf {
  id: ID!
  dimensions: ProductDimension
  sku: String
  name: String
  package: String
  variation: ProductVariation
  createdBy: User
  hidden: String @inaccessible
  oldField: String @deprecated(reason: "refactored out")
}

type ProductVariation {
  id: ID!
  name: String
}

type Query {
  allProducts: [ProductItf]
  product(id: ID!): ProductItf
  _entities(representations: [_Any!]!): _Entity
}

enum ShippingClass {
  STANDARD
  EXPRESS
  OVERNIGHT
}

interface SkuItf {
  sku: String
}

type User {
  email: ID! @tag(name: "test-from-users")
  totalProductsCreated: Int
}

union _Entity = Product | User

scalar _Any"
`;

exports[`Supergraphs b.graphql subgraphs: b.graphql - REVIEWS 1`] = `
"schema {
  query: Query
}

type Product implements ProductItf {
  id: ID! @tag(name: "hi-from-products")
  reviewsScore: Float!
  reviewsCount: Int!
  reviews: [Review!]!
}

interface ProductItf {
  id: ID!
  reviewsCount: Int!
  reviewsScore: Float!
  reviews: [Review!]!
}

type Query {
  review(id: Int!): Review
  _entities(representations: [_Any!]!): _Entity
}

type Review {
  id: Int!
  body: String!
}

union _Entity = Product

scalar _Any"
`;

exports[`Supergraphs b.graphql subgraphs: b.graphql - USERS 1`] = `
"schema {
  query: Query
}

type Query {
  _entities(representations: [_Any!]!): _Entity
}

type User {
  email: ID! @tag(name: "test-from-users")
  totalProductsCreated: Int
  name: String
}

union _Entity = User

scalar _Any"
`;

exports[`Supergraphs c.graphql matches: c.graphql - stitchedSchema 1`] = `
"schema {
  query: Query
}

type Query {
  me: User
  throw: String
  fooBar: FooBar
  mustFooBar: FooBar!
}

type User {
  id: ID!
  username: String
}

union FooBar = Foo | Bar

type Foo {
  id: ID!
  name: String
}

type Bar {
  id: ID!
  name: String
}"
`;

exports[`Supergraphs c.graphql subgraphs: c.graphql - ACCOUNTS 1`] = `
"schema {
  query: Query
}

type Query {
  me: User
  throw: String
  fooBar: FooBar
  mustFooBar: FooBar!
  _entities(representations: [_Any!]!): _Entity
}

type User {
  id: ID!
  username: String
}

union FooBar = Foo | Bar

type Foo {
  id: ID!
  name: String
}

type Bar {
  id: ID!
  name: String
}

union _Entity = User | Foo | Bar

scalar _Any"
`;
